TaskFramework.md

```diff
@@ -0,0 +1,5 @@
+# Task Framework
+I want you to design a task framework. This framework will trigger 1 billion tasks everyday. A task can be defined as a code executable. We want our task framework to support task runners in all modern programming languages. This task framework also needs to have a topological ordering, in which we can trigger subtasks and the sub tasks can trigger subtasks. Some tasks may be dependent on specific things but that is up for the user of our framework to implement. 
```

Comment:
1. What's a code executable? Like running a binary? 
2. Assuming that tasks can have subtasks and those subtasks can have subtasks infinitely, we'll have to fulfill our dependencies before we run the task at hand. Are there any other things to keep in mind other than subtasks, like requiring linked libraries for tasks? (Systemd allows you to require that some service is started or a library exists before running a binary, like C libraries require a libc before being run, since they're normally dynamically linked).
3. How heavy are these tasks? How much hardware do we have? We need to run about 10k tasks per second, (1B / 10k seconds in a day). 
4. Should I be providing some sort of interface that allows the users to ask for their dependencies up front like systemd does, or is that out of scope?
5. Do I need to care about analytics and logging?
6. Can I assume that each task runs on one machine, or might tasks run on many?
7. What should I do if a task fails? Should I retry it in a bit, or can I bubble up an error to the user?

Reply:
1. Could be a binary could be a python file. We also may want to suport some yaml based system where we can trigger one executable after another in sequential order or in tandem.

Reply:
2. yes dependency recovery is a real thing but keep in mind it falls on the user of the framework to provide code with valid libs.

Reply:
3. This can depend we will probably need some sort of auto provisioning hardware to manage it. 
4. dependencies are pretty straightforward in most languages, lets say there is another team at bry-takashi-inc that is responsible for building a tool to pull dependencies from the file, but our system will still be responsible for installing those extracted dependencies

Reply:
5. YES of course we are big data organization

Reply:
6. Thats up to you and most likely depends on the taski

Reply:
7. Retry in intervals each task will be configured with a number of retries